---
title: "BCB420-A1"
output:
  html_document:
    df_print: paged
---
## 1. Preparations:

#Packages
load all the packages required
```{r}
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")
if (!requireNamespace("GEOmetadb", quietly = TRUE))
  BiocManager::install("GEOmetadb")
if(!requireNamespace("edgeR", quietly = TRUE))
  BiocManager::install("edgeR")
if(!requireNamespace("biomaRt", quietly = TRUE))
  BiocManager::install("biomaRt")
if(!requireNamespace("ggplot2", quietly = TRUE))
  BiocManager::install("ggplot2")
if(!requireNamespace("reshape2", quietly = TRUE))
  BiocManager::install("reshape2")

library("BiocManager")
library("GEOmetadb")
library("edgeR")
library("biomaRt")
library("knitr")
library("ggplot2")
library("reshape2")
```

## 2. Data exploration:
#The datasets choosed is GSE150621. The following step is data exploration of the dataset chose. 
```{r}
sfiles <- getGEOSuppFiles('GSE150621')
fnames <- rownames(sfiles)
exp_data = read.delim(fnames[1],header=TRUE, check.names = FALSE)  #load the counts data into R 
kable(exp_data, format = "html") #show the first 15 rows and 5 colomns of b2
```
# Explore the dataset
```{r}
gse <- getGEO("GSE150621")
gsms <- as.data.frame(pData(gse[[1]]))
gsms
gsms[,1]
gsms$characteristics_ch1 #There are 8 controls and 6 GDM exposure groups
gsms$characteristics_ch1.1 #7 males and 7 females
gsms$characteristics_ch1.2 # There are 11 samples' gestational age of amniocentesis are 16 weeks, and 17 weeks for the rest 3 samples. 
gsms$characteristics_ch1.3 # maternal age occurrences: 34(1),35(4), 36(2), 38(2), 39(1), 40(1), 41(1), 42(1), 44(1)
gsms$characteristics_ch1.4 # gestational age birth occurrences: 38(5), 39(6), 40(1), 41(2)
```


## 3. Data cleaning 

#What is the dimension of our data and the colomn names of the data?
```{r}
dim(exp_data)
colnames(exp_data) #There are 14 different groups in the dataset. Their series files show that the last two letters determine the exposure(control/gdm), and the sex of the group. "C" refers to control, "D" refers to gdm exposure group, "F" refers to female, and "M" refers to male. There are 8 controls groups and  6 gdm exposure groups. 7 groups for females, and 7 groups for males. 
```
#Group them by their patient ids, conditions, and patient sex
```{r}
patient_id <- substring(colnames(exp_data[2:ncol(exp_data)]), 8, 10)
condition <- ifelse(substring(colnames(exp_data[2:ncol(exp_data)]), 12, 12) == "C", "control", "DPM")
sex <- ifelse(substring(colnames(exp_data[2:ncol(exp_data)]), 13, 13) == "F", "female", "male")
samples <- data.frame(Patient = patient_id, Condition = condition, Sex = sex)
rownames(samples)=colnames(exp_data[2:ncol(exp_data)])
samples
```

#check for the duplicated genes 
```{r}
summarized_gene_counts <- sort(table(exp_data$refGene), decreasing = TRUE)
kable(summarized_gene_counts[which(summarized_gene_counts>1)[1:10]], format="html")
#It shows that entries "1-Mar" and "2-Mar" has a frequency of 2
a = subset(exp_data, refGene == "1-Mar")
b = subset(exp_data, refGene == "2-Mar")
a
b
#It shows that "1-Mar" and "2-Mar" have values in groups. It might be some man made errors when manipulating data.  Therefore, they should be removed in order to keep the accuracy for the later downsteam analysis. 
```


#According to the edgeR protocol, filter weakly expressed and noninformative data.
```{r}
cpms = cpm(exp_data[, 2:15])
rownames(cpms) <- exp_data[,1]
keep = rowSums(cpms >1) >= 6  #We have 8 controls and 6 DPM exposure groups. Therefore, it should be 6 which is the smallest number of replicates in groups.
data_filtered = exp_data[keep,]
```
#now check the dimension of the filtered dataset.
```{r}
dim(data_filtered) #the filtered expression data numbers
```
#check whether if it solves some duplicate issues
```{r}
summarized_gene_counts_filtered <- sort(table(data_filtered$refGene),
decreasing = TRUE)
kable(summarized_gene_counts_filtered[
which(summarized_gene_counts_filtered>1)[1:10]],
format="html")
```
#Remove the duplicates gene that shows invalid gene names "1-Mar" and "2-Mar" .\
```{r}
data_filtered <- data_filtered[!(data_filtered[,1] %in% c("1-Mar", "2-Mar")), ]
dim(data_filtered) #line 116 returns 14149 rows, and this means 4 invalid rows contains "1-Mar"or "2-Mar" as refGene are successfully removed here. 

```
#Mapping: most of the genes look like HGNC symbols. Check for whether all of them are. 
```{r}
# Load the ensemble human gene database
ensembl <- useMart("ensembl", dataset = "hsapiens_gene_ensembl")

# Get the gene symbol and HGNC symbol mappings
gene_symbol_mapping <- getBM(attributes = c("hgnc_symbol", "hgnc_id"), mart = ensembl)

# Create a vector to store the HGNC gene symbols or "NA"
hgnc_gene_identifiers <- character(nrow(data_filtered))

# Loop through each row of the data_filtered list
for (i in 1:nrow(data_filtered)) {
  gene_symbol <- data_filtered[i, 1]
  
  # Check if the gene symbol exists in the gene_symbol_mapping data frame
  symbol_exists <- gene_symbol %in% gene_symbol_mapping$hgnc_symbol
  
  if (symbol_exists) {
    hgnc_gene_identifiers[i] <- gene_symbol
  } else {
    hgnc_gene_identifiers[i] <- "NA"
  }
}

# Add the HGNC gene symbols or "NA" to the "HGNC genes" column
data_filtered <- cbind(data_filtered[, 1], "HGNC genes" = hgnc_gene_identifiers, data_filtered[, -1])
colnames(data_filtered)[1] <- "refGene"

num_na <- sum(data_filtered[, 2] == "NA")

# Print the result
print(paste("Number of NA values in the HGNC genes column:", num_na))
kable(data_filtered, format = "html")


```

```{r}
# Find the row indices of the "NA" values in the "HGNC_symbol" column
na_rows <- which(data_filtered[, 2]==
                   "NA")

# Extract the corresponding entries from the first column (refGene)
gene_identifiers <- data_filtered[na_rows, 1]

# Write the gene identifiers to a text file
write.table(gene_identifiers, file = "gene_identifiers_without_HGNC.txt", row.names = FALSE, col.names = FALSE)

```


```{r}
# Read the content of the text file line by line
lines = readLines("gene_identifiers_without_HGNC.txt")

# Count the number of lines that start with "LOC"
loc_lines = grep("LOC", lines)

# Show the result
cat("The number of lines that start with 'LOC' is:", length(loc_lines)) #about 342/1080 = 0.3 identifiers in file start with "LOC"  are non coding RNA 

```


#Pre-normalized charts:
#Box plot
```{r}
data2plot <- log2(cpm(data_filtered[, 3:ncol(data_filtered)]))
data2plot <- as.data.frame(data2plot)
colnames(data2plot) <- colnames(data_filtered[, 3:ncol(data_filtered)])

data2plot_melted <- melt(data2plot)
ggplot(data2plot_melted, aes(x = variable, y = value)) +
  geom_boxplot() +
  geom_hline(yintercept = median(data2plot_melted$value), color = "green", linetype = "dashed") +
  labs(x = "Samples", y = "log2 CPM") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  ggtitle("DPM RNASeq Samples")
```

#Density Plot
```{r}
counts_density <- 
  apply(log2(cpm(data_filtered[, 3:length(colnames(data_filtered))])),
        2, density)
 #calculate the limits across all the samples
xlim <- 0; ylim <- 0
for (i in 1:length(counts_density)) {
  xlim <- range(c(xlim, counts_density[[i]]$x)); 
  ylim <- range(c(ylim, counts_density[[i]]$y))
}
 cols <- rainbow(length(counts_density))
 ltys <- rep(1, length(counts_density))
 
 #plot the first density plot to initialize the plot
 plot(counts_density[[1]], 
      xlim=xlim, 
      ylim=ylim, 
      type="n", 
      ylab="Smoothing density of log2-CPM", 
      main="Density Plot", 
      cex.lab = 0.8)
 
 #plot each line
 for (i in 1:length(counts_density)) {
   lines(counts_density[[i]], 
         col = cols[i], 
         lty = ltys[i])
 }
   
 #create legend
 legend("topright", colnames(data2plot), 
 col=cols, lty=ltys, cex=0.75, 
 border ="blue", text.col = "green4", 
 merge = TRUE, bg = "gray90")
```

## 4. Normalization

#creating our DGEList objects to be used by edgeR
```{r}
filter_data_matrix <- as.matrix(data_filtered[,3:ncol(data_filtered)])
rownames(filter_data_matrix) <- data_filtered$refGene
d = DGEList(counts=filter_data_matrix,group=samples$Condition)
```

#Calculate the normalization factors
```{r}
d = calcNormFactors(d)
normalized_counts <- cpm(d)
```

#Inspect the sample separation using a multidimenstional scaling plot.
```{r}
plotMDS(d, labels=rownames(samples),
        col = c("darkgreen","blue")[factor(samples$Condition)])
```
#Normalized box plot
```{r}
data2plot <- log2(cpm(normalized_counts[, 3:ncol(normalized_counts)]))
data2plot <- as.data.frame(data2plot)
colnames(data2plot) <- colnames(normalized_counts[, 3:ncol(normalized_counts)])

data2plot_melted <- melt(data2plot)
ggplot(data2plot_melted, aes(x = variable, y = value)) +
  geom_boxplot() +
  geom_hline(yintercept = median(data2plot_melted$value), color = "green", linetype = "dashed") +
  labs(x = "Samples", y = "log2 CPM") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  ggtitle("DPM RNASeq Samples")
```
#Normalized Density Plot
```{r}
counts_density <- 
  apply(log2(cpm(normalized_counts[, 3:length(colnames(normalized_counts))])),
        2, density)
 #calculate the limits across all the samples
xlim <- 0; ylim <- 0
for (i in 1:length(counts_density)) {
  xlim <- range(c(xlim, counts_density[[i]]$x)); 
  ylim <- range(c(ylim, counts_density[[i]]$y))
}
 cols <- rainbow(length(counts_density))
 ltys <- rep(1, length(counts_density))
 
 #plot the first density plot to initialize the plot
 plot(counts_density[[1]], 
      xlim=xlim, 
      ylim=ylim, 
      type="n", 
      ylab="Smoothing density of log2-CPM", 
      main="Density Plot", 
      cex.lab = 0.8)
 
 #plot each line
 for (i in 1:length(counts_density)) {
   lines(counts_density[[i]], 
         col = cols[i], 
         lty = ltys[i])
 }
   
 #create legend
 legend("topright", colnames(data2plot), 
 col=cols, lty=ltys, cex=0.75, 
 border ="blue", text.col = "green4", 
 merge = TRUE, bg = "gray90")
```
#Dispersion
```{r}
model_design <- model.matrix(~samples$Condition+0)
d <- estimateDisp(d, model_design)
plotBCV(d,col.tagwise = "black",col.common = "red",)
```
```{r}
plotBCV(d,col.tagwise = "black",col.common = "red",)
```
```{r}
plotMeanVar(d, show.raw.vars = TRUE, 
            show.tagwise.vars=TRUE, 
            show.ave.raw.vars = TRUE, 
            NBline=TRUE,
            show.binned.common.disp.vars = TRUE)
```


##Interpretation
1. What are the control and test conditions of the dataset?
Controls are amniocytes not exposed to Gestational diabetes (GDM).
Test conditions are amniocytes exposed to Gestational diabetes (GDM).
A nested case-control study was performed in second trimeseter amniocytes matched for offspring sex, maternal race/ethnicity, maternal age, gestational age at amniocentesis, gestational age at birth and gestational diabetes status.

2. Why is the dataset of interest to you?
My grandmother was diabetic and this led to almost all of her offspring having diabetes. I am therefore very interested in the effects of gestational diabetes on offsprings. Alterations gene expression in amniocytes exposed to GDM in utero may identify potential mechanisms leading to metabolic dysfunction later in life.

3. Were there expression values that were not unique for specific genes? How did you handle these?
In the processing of checking duplicated genes, I found two entries with refGene "Mar-1" and "Mar-2" that has no relation to any gene representations, and they both appear twice in the dataset. Therefore, I plan to remove them since they might be man-made errors in the process and they do not contain any informative data that could be used for the later downstream analysis. 

4. Were there expression values that could not be mapped to current HUGO symbols?
Most of the refGenes are HUGO symbols. But there are about 1080 rows that have gene names that could not be mapped, some of them are outdated aliases of HUGO symbols, some of them are non coding RNAs. 

5. How many outliers were removed?
The original dataset contains 27135 genes, and 12801 were removed in filtering weak expression values, and 4 were removed in checking duplicated genes. 

6. How did you handle replicates?
There are 6 replicates of GDM exposured group and 8 replicates of the controls. I grouped them into "Condition" and "Sex" based on the last two letters of their datasets. 

7. What is the final coverage of your dataset?
14330 expression values and 14 samples in my final dataset. 


##reference
Lecture modules:https://q.utoronto.ca/courses/248455/files/19147947?module_item_id=3463364

Ke-Ying Fang, Zi-Qi Liu, Qi-Lin Hu et al. Metabolic Effect and Mechanism of Gestational Diabetes Mellitus on Offspring of Different Sexes, 30 September 2021, PREPRINT (Version 1) available at Research Square [https://doi.org/10.21203/rs.3.rs-947777/v1]



